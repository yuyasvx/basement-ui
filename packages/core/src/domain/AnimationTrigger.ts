export const AnimationTrigger = {
  SHOW: 'show',
  HIDE: 'hide',
  BOTH: 'both',
  NONE: 'none'
} as const;
