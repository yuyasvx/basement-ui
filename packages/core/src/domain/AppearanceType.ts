import { Case } from '../util/Case';

export const AppearanceType = {
  NORMAL: 'normal',
  TINT: 'tint',
  FLAT: 'flat',
  SUPER_FLAT: 'super_flat',
  MINIMAL: 'minimal'
} as const;

const appearanceMap: Record<string, string> = {
  [AppearanceType.NORMAL]: '--normal',
  [AppearanceType.TINT]: '--tint',
  [AppearanceType.FLAT]: '--flat',
  [AppearanceType.SUPER_FLAT]: '--super-flat'
};

export const getAppearanceClassName = (appearance?: Case<typeof AppearanceType>): Case<typeof appearanceMap> => {
  return appearance != null ? appearanceMap[appearance] : appearanceMap[AppearanceType.NORMAL];
};
