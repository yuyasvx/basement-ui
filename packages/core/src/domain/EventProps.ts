import { SyntheticEvent, ChangeEvent, MouseEventHandler, FormEventHandler, KeyboardEventHandler } from 'react';

export interface MouseEvents<EL = Element> {
  onClick?: MouseEventHandler<EL>;
  onMouseOver?: MouseEventHandler<EL>;
  onDoubleClick?: MouseEventHandler<EL>;
  onContextMenu?: MouseEventHandler<EL>;
  onMouseDown?: MouseEventHandler<EL>;
  onMouseUp?: MouseEventHandler<EL>;
  onDrag?: MouseEventHandler<EL>;
  onDragStart?: MouseEventHandler<EL>;
  onDragEnd?: MouseEventHandler<EL>;
  onDragEnter?: MouseEventHandler<EL>;
  onDragExit?: MouseEventHandler<EL>;
  onDragLeave?: MouseEventHandler<EL>;
  onDragOver?: MouseEventHandler<EL>;
  onDrop?: MouseEventHandler<EL>;
  onMouseEnter?: MouseEventHandler<EL>;
  onMouseLeave?: MouseEventHandler<EL>;
  onMouseMove?: MouseEventHandler<EL>;
  onMouseOut?: MouseEventHandler<EL>;
}

export interface KeyEvents<EL = Element> {
  onKeyDown?: KeyboardEventHandler<EL>;
  onKeyDownCapture?: KeyboardEventHandler<EL>;
  onKeyPress?: KeyboardEventHandler<EL>;
  onKeyPressCapture?: KeyboardEventHandler<EL>;
  onKeyUp?: KeyboardEventHandler<EL>;
  onKeyUpCapture?: KeyboardEventHandler<EL>;
}

export interface FocusEvents<EL = Element> {
  onFocus?: <EL, EV = Event>(event: SyntheticEvent<EL, EV>) => void;
  onBlur?: <EL, EV = Event>(event: SyntheticEvent<EL, EV>) => void;
}

export interface FormEvents<EL = Element> {
  onChange?: (evt: ChangeEvent<EL>) => void;
  onInput?: FormEventHandler<EL>;
  onInvalid?: FormEventHandler<EL>;
  onReset?: FormEventHandler<EL>;
  onSubmit?: FormEventHandler<EL>;
}

export interface PointerEvents<EL = Element> {
  onPointerDown?: <EL, EV = Event>(event: SyntheticEvent<EL, EV>) => void;
  onPointerMove?: <EL, EV = Event>(event: SyntheticEvent<EL, EV>) => void;
  onPointerUp?: <EL, EV = Event>(event: SyntheticEvent<EL, EV>) => void;
  onPointerCancel?: <EL, EV = Event>(event: SyntheticEvent<EL, EV>) => void;
  onGotPointerCapture?: <EL, EV = Event>(event: SyntheticEvent<EL, EV>) => void;
  onLostPointerCapture?: <EL, EV = Event>(event: SyntheticEvent<EL, EV>) => void;
  onPointerEnter?: <EL, EV = Event>(event: SyntheticEvent<EL, EV>) => void;
  onPointerLeave?: <EL, EV = Event>(event: SyntheticEvent<EL, EV>) => void;
  onPointerOver?: <EL, EV = Event>(event: SyntheticEvent<EL, EV>) => void;
  onPointerOut?: <EL, EV = Event>(event: SyntheticEvent<EL, EV>) => void;
}
