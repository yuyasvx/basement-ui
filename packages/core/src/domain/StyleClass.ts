// TODO 謎の整理の仕方
export const RootStyle = {
  BASE: 'bm-base',
  TEXT_BASE: 'bm-text-base',
  CONTENT_BASE: 'bm-content-base'
} as const;

export const PUSHABLE_STYLE = 'bm-a-pushable';

export const FOCUSABLE_STYLE = 'bm-a-focusable';

export const backgroundStyle = ['bm-a-bg-lv1', 'bm-a-bg-lv2', 'bm-a-bg-lv3', 'bm-a-bg-lv4'];
export const shadowStyle = ['bm-a-shadow-lv1', 'bm-a-shadow-lv2', 'bm-a-shadow-lv3', 'bm-a-shadow-lv4'];
export const blurStyle = ['bm-a-blur-lv1'];
