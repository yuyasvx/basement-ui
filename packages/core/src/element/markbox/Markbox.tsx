import clsx from 'clsx';
import { FC, ForwardedRef, PropsWithChildren, ReactNode, forwardRef, useMemo } from 'react';
import { BaseComponentProps, getBaseComponentProps } from '../../base/BaseComponent';
import { StyleElementProps, StyleElements, useStyleElement } from '../../style-element/StyleElementHook';

const NAME = 'bm-e-markbox';

export type MarkboxProps = {
  marked: boolean;
  mark?: ReactNode;
} & StyleElementProps &
  BaseComponentProps;

export const Markbox = forwardRef((props: MarkboxProps, ref: ForwardedRef<HTMLDivElement>) => {
  const { newProps, symbolProps } = useMarkboxElement(props);
  return (
    <div ref={ref} {...newProps}>
      <MarkboxSymbolContainer {...symbolProps}>{props.mark}</MarkboxSymbolContainer>
    </div>
  );
});

export const MarkboxSymbolContainer: FC<PropsWithChildren<BaseComponentProps>> = (props) => {
  return <div className={props.className}>{props.children}</div>;
};

export const useMarkboxElement = (props: MarkboxProps) => {
  const styleElement = useStyleElement(StyleElements.PUSH, {
    variant: props.variant,
    effect: props.effect
  });
  const baseProps = getBaseComponentProps(props);
  return {
    name: NAME,
    newProps: {
      className: useMemo(
        () =>
          clsx(NAME, styleElement.name, styleElement.variant, styleElement.manual, styleElement.manualEffect, {
            '-marked': props.marked
          }),
        [props.marked, styleElement.manual, styleElement.manualEffect, styleElement.name, styleElement.variant]
      ),
      ...baseProps
    },
    symbolProps: {
      className: `${NAME}__symbol`
    }
  };
};
