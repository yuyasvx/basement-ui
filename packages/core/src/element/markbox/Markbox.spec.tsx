import { cleanup, render, renderHook } from '@testing-library/react';
import { AppearanceType } from '../../domain/AppearanceType';
import { Markbox, useMarkboxElement } from './Markbox';

describe('Markbox Element Snapshot', () => {
  afterEach(() => {
    cleanup();
  });

  test('必須Propsのみ', () => {
    const { asFragment } = render(<Markbox mark={'X'} marked={false} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('Props', () => {
    const { asFragment } = render(<Markbox mark={'X'} marked variant={AppearanceType.TINT} effect={'disabled'} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('Base Props', () => {
    const { asFragment } = render(
      <Markbox
        mark={'-'}
        marked
        id={'example'}
        tabIndex={2}
        style={{ width: '100px' }}
        className="example-class"
        nativeProps={{ 'data-test': true }}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('useMarkboxElement', () => {
  test('Style Elementが適用されている', () => {
    const rendered = renderHook(() => useMarkboxElement({ variant: AppearanceType.NORMAL, mark: 'X', marked: true }));
    expect(rendered.result.current.newProps.className).toBe('bm-e-markbox bm-s-push --normal -marked');
  });

  test('Markを解除するとclassNameに反映される', () => {
    const rendered = renderHook(() => useMarkboxElement({ variant: AppearanceType.NORMAL, mark: 'X', marked: false }));
    expect(rendered.result.current.newProps.className).toBe('bm-e-markbox bm-s-push --normal');
  });
});
