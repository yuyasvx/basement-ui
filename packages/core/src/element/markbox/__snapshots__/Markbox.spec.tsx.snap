// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markbox Element Snapshot Base Props 1`] = `
<DocumentFragment>
  <div
    class="bm-e-markbox bm-s-push --normal -marked"
    data-test="true"
    id="example"
    style="width: 100px;"
    tabindex="2"
  >
    <div
      class="bm-e-markbox__symbol"
    >
      -
    </div>
  </div>
</DocumentFragment>
`;

exports[`Markbox Element Snapshot Props 1`] = `
<DocumentFragment>
  <div
    class="bm-e-markbox bm-s-push --tint ---manual ---disabled -marked"
  >
    <div
      class="bm-e-markbox__symbol"
    >
      X
    </div>
  </div>
</DocumentFragment>
`;

exports[`Markbox Element Snapshot 必須Propsのみ 1`] = `
<DocumentFragment>
  <div
    class="bm-e-markbox bm-s-push --normal"
  >
    <div
      class="bm-e-markbox__symbol"
    >
      X
    </div>
  </div>
</DocumentFragment>
`;
